### DISCUSSION TITLE
Create HSL color and convert it to Hex RGB string

### DISCUSSION THREAD

**Discussion by BlueskyFR** [Ogeon/palette#359] [Q&A] [answered]
Hey! I am new to Rust & to this crate and I can't find how to easily create a HSL color and then print it as an RGB hex (`#ff0000`) string.

Maybe it's just me but I feel like the documentation is super hard to understand when manipulating colors should be pretty straightforward, but I guess I am just missing something.

Thanks in advance!

**Comment by Ogeon** [comment, marked as answer]
Hello! All the bits and pieces are there, but maybe not all put front and center. I would like to improve that. Until that happens, here's a recipe for HSL to HEX RGB:

```rust
use palette::{Hsl, Srgb, FromColor};

// Assuming f32 input, but depends on your case.
let hsl = Hsl::new(123.0f32, 0.5, 0.8);

// Convert to RGB and then convert the float components to u8
// The type of RGB is perhaps not important in your case, but sRGB is the typical output.
let rgb_u8: Srgb<u8> = Srgb::from_color(hsl).into_format();

// The hex conversion is a formatting option. The `:x` means hexadecimal formatting.
// The # is not a mandatory part of the RGB format, so has to be added by us.
let rgb_hex = format!("#{rgb_u8:x}");
```

I didn't test run this, so let me know if there are any mistakes. Here's also a preview of an updated `Rgb` documentation that may help as well: https://ogeon.github.io/palette/palette/rgb/struct.Rgb.html

**Reply by BlueskyFR** [reply]
Thanks! The library looks really interesting since it has a lot of options, but for cases like this it is a bit hard to handle without being familiar with it.
Would something like `format!("#{}", Hsl::new(123.0_f32, 0.5, 0.8).to_hex())` be possible in the future?

What I mean is that the user doesn't have to know/learn that conversion to u8 has to be done for instance IMO

Thanks for your help!

**Reply by Ogeon** [reply]
Maybe. Part of the library's design is to be open to all those different options, so there are at least a couple of assumptions it avoids:
 * It doesn't assume that you want 8 bits per channel ([could be 16 bits, like in a lot of X11 stuff](https://superuser.com/questions/122114/48-bit-colours)).
 * It doesn't assume that you want to allocate a string just for the color, so it uses the formatting system that's more lazy and universal.

A lot of color libraries simplify things or aim for what you usually need for the web, so they can take a lot of shortcuts. Palette is taking the possibility of innumerable RGB standards into account, as well as more esoteric color spaces as source data. 😰 Most of the work goes into providing the building blocks. Colors are one of those fields, like text/strings, that have been made to look more simple than they are.

That said, shortcuts like this could be added, and have been added in the past. I just haven't considered this particular one so far.

**Reply by BlueskyFR** [reply]
Sure! I chose Palette because of the diversity of options it provides but I believe a higher-level API like .to_hex() can co-exist with the current one, so that usage is greatly simplified but the users would still have the choice of having more control over what is happening
Thanks for your work!

**Reply by Ogeon** [reply]
It sure can, it just hasn't been considered or prioritized yet.

**Reply by Ogeon** [reply]
...and thanks for the kind words and for taking the time to ask. It shows that there's still room for improvements in this area.
